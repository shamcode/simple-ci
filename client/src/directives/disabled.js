export default class {
    constructor() {
        this.node = null;
    }

    bind( node ) {
        this.node = node;
    }

    unbind() {
        this.node = null;
    }

    update( value ) {
        this.node.disabled = value;
    }
}
