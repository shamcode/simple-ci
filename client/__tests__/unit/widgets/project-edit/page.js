import { DI } from 'sham-ui';
import Page from '../../../../src/widgets/project-edit/page';
import disabled from '../../../../src/directives/disabled';
import directives from '../../../../src/directives/event-listener';
import renderer from 'sham-ui-test-helpers';

const flushPromises = () => new Promise( resolve => setImmediate( resolve ) );

afterEach( () => {
    DI.bind( 'store', null );
    DI.bind( 'router', null );
    jest.resetModules();
    jest.clearAllMocks();
} );

it( 'renders correctly', async () => {
    expect.assertions( 5 );

    const storeMock = jest.fn().mockReturnValue( Promise.resolve( {
        project: {
            id: 1,
            name: 'Test',
            cwd: '/tmp/'
        }
    } ) );
    DI.bind( 'store', {
        getProjectById: storeMock
    } );
    DI.bind( 'router', {
        lastRouteResolved: jest.fn().mockReturnValueOnce( { params: { id: 1 } } ),
        generate: jest.fn().mockReturnValueOnce( '/' )
    } );

    const meta = renderer( Page, {
        directives: {
            disabled,
            ...directives
        }
    } );
    const { widget: { container } } = meta;
    expect( storeMock.mock.calls[ 0 ][ 0 ] ).toBe( 1 );
    expect( meta.toJSON() ).toMatchSnapshot();

    await flushPromises();
    expect( container.querySelector( '[name="name"]' ).value ).toBe( 'Test' );
    expect( container.querySelector( '[name="cwd"]' ).value ).toBe( '/tmp/' );
    expect( meta.toJSON() ).toMatchSnapshot();
    await flushPromises();
} );

it( 'update fail', async() => {
    expect.assertions( 2 );

    const updateProjectMock = jest.fn();
    DI.bind( 'store', {
        getProjectById: jest.fn().mockReturnValue( Promise.resolve( {
            project: {
                id: 1,
                name: 'Test',
                cwd: '/tmp/'
            }
        } ) ),
        updateProject: updateProjectMock
    } );
    DI.bind( 'router', {
        lastRouteResolved: jest.fn().mockReturnValueOnce( { params: { id: 1 } } ),
        generate: jest.fn().mockReturnValueOnce( '/' )
    } );

    const meta = renderer( Page, {
        directives: {
            disabled,
            ...directives
        }
    } );
    const { widget: { container } } = meta;
    await flushPromises();


    updateProjectMock.mockReturnValueOnce( Promise.reject() );
    container.querySelector( '[type="submit"]' ).click();
    expect( updateProjectMock.mock.calls.length ).toBe( 1 );

    await flushPromises();
    expect( meta.toJSON() ).toMatchSnapshot();
} );