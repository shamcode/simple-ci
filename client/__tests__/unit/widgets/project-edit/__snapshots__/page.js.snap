// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Object {
  "ID": "widget",
  "constructorOptions": Object {
    "directives": Object {
      "disabled": [Function],
      "onclick": [Function],
      "onsubmit": [Function],
    },
  },
  "html": "
<!--if-->
<!--FlushPageDataLoading-->
<!--if-->
",
  "nodes": Array [
    Object {
      "ID": undefined,
      "constructorOptions": null,
      "html": null,
      "nodes": null,
      "options": null,
    },
  ],
  "options": Object {
    "actionSequence": Array [
      "render",
      "bindEvents",
    ],
    "afterRender": [Function],
    "dataLoaded": false,
    "dataSaving": false,
    "directives": Object {
      "disabled": [Function],
      "onclick": [Function],
      "onsubmit": [Function],
    },
    "errors": Array [],
    "project": Object {},
    "renderAsync": false,
    "renderDependence": Array [],
    "types": Array [],
  },
}
`;

exports[`renders correctly 2`] = `
Object {
  "ID": "widget",
  "constructorOptions": Object {
    "directives": Object {
      "disabled": [Function],
      "onclick": [Function],
      "onsubmit": [Function],
    },
  },
  "html": "
<!--if-->
<!--ErrorsPanel-->
<div class=\\"project-create-form\\">
  <form>
    <div class=\\"group\\">
      <input name=\\"name\\" type=\\"text\\" required=\\"\\"><span class=\\"highlight\\"></span><span class=\\"bar\\"></span>
      <label>Name</label>
    </div>
    <div class=\\"group\\">
      <input name=\\"cwd\\" type=\\"text\\" required=\\"\\"><span class=\\"highlight\\"></span><span class=\\"bar\\"></span>
      <label>Current Work Directory</label>
    </div>
    <div class=\\"group\\">
      <button type=\\"submit\\"><span>Update</span></button>
    </div>
  </form>
</div>
<!--ProjectForm-->
<!--if-->
",
  "nodes": Array [
    Object {
      "ID": undefined,
      "constructorOptions": null,
      "html": null,
      "nodes": null,
      "options": null,
    },
  ],
  "options": Object {
    "actionSequence": Array [
      "render",
      "bindEvents",
    ],
    "afterRender": [Function],
    "dataLoaded": true,
    "dataSaving": false,
    "directives": Object {
      "disabled": [Function],
      "onclick": [Function],
      "onsubmit": [Function],
    },
    "errors": Array [],
    "project": Object {
      "cwd": "/tmp/",
      "id": 1,
      "name": "Test",
    },
    "renderAsync": false,
    "renderDependence": Array [],
    "types": Array [],
  },
}
`;

exports[`update fail 1`] = `
Object {
  "ID": "widget",
  "constructorOptions": Object {
    "directives": Object {
      "disabled": [Function],
      "onclick": [Function],
      "onsubmit": [Function],
    },
  },
  "html": "
<ul class=\\"errors-panel\\">
  <li>Update project fail!</li>
</ul>
<!--if-->
<!--ErrorsPanel-->
<div class=\\"project-create-form\\">
  <form>
    <div class=\\"group\\">
      <input name=\\"name\\" type=\\"text\\" required=\\"\\"><span class=\\"highlight\\"></span><span class=\\"bar\\"></span>
      <label>Name</label>
    </div>
    <div class=\\"group\\">
      <input name=\\"cwd\\" type=\\"text\\" required=\\"\\"><span class=\\"highlight\\"></span><span class=\\"bar\\"></span>
      <label>Current Work Directory</label>
    </div>
    <div class=\\"group\\">
      <button type=\\"submit\\"><span>Update</span></button>
    </div>
  </form>
</div>
<!--ProjectForm-->
<!--if-->
",
  "nodes": Array [
    Object {
      "ID": undefined,
      "constructorOptions": null,
      "html": null,
      "nodes": null,
      "options": null,
    },
  ],
  "options": Object {
    "actionSequence": Array [
      "render",
      "bindEvents",
    ],
    "afterRender": [Function],
    "dataLoaded": true,
    "dataSaving": false,
    "directives": Object {
      "disabled": [Function],
      "onclick": [Function],
      "onsubmit": [Function],
    },
    "errors": Array [
      "Update project fail!",
    ],
    "project": Object {
      "cwd": "/tmp/",
      "id": 1,
      "name": "Test",
    },
    "renderAsync": false,
    "renderDependence": Array [],
    "types": Array [],
  },
}
`;
